### ISSUE TITLE
Add { Insert | Update | Delete }ReturningStep.returningRecord(...) methods

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#7475]
A historic API design flaw prevents users from leveraging the full power of the `{ INSERT | UPDATE | DELETE } .. RETURNING` clause, which allows for projecting arbitrary column expressions in the `RETURNING` clause, not just actual columns in the table being updated (#5070).

For backwards compatibility reasons, we cannot easily change the API, but we can introduce new API that behaves as expected. jOOQ 4.0 will then deprecate the new API again, changing the old one to the expected behaviour.

We'll take the opportunity to overload the new API with the usual `Record[N]` types.

**Comment by JanecekPetr** [comment]
> It will be desireable for that method to return a `ResultQuery` subtype instead of (or in addition to) `{ Insert | Update | Delete }ResultStep`. This way, we can have all sorts of additional `fetch()` methods that have been requested many times for RETURNING statements.

I got so excited for a moment here! For any future readers, this unfortunately *did not make it into the final implementation*, so we're stuck with only `fetch()`, `fetchOptional()` and `fetchOne()` even with the new `.returningResult()` call, too. Apparently returning queries are hard.

Should we look forward to jOOQ 4?

**Comment by lukaseder** [comment]
Thank you very much for your comment. This isn't about the problem being hard, but about API design strategies.

The important feature here was to be able to project arbitrary column expressions in a `RETURNING` clause. The far less important feature (even if useful) is to have additional convenience on that particular query type.

- Should we let the resulting query extend `ResultQuery`? (That raises many questions about many internals, emulations, and other features that currently interact with `ResultQuery`. It would have been an extremely buggy implementation)
- Should we copy all the `ResultQuery` API *again* to all the `ReturningStep` types? That is hardly maintainable. We'd forget some methods in the future.
- Should we explore something new? We will be, including moving all the existing `fetch` convenience into a dedicated `Collector` API (see `ResultQuery.collect`), which will compose much better, do exactly what people want (collect and map data to other data types), without any of the implicit implications of pretending that these query types are indeed `ResultQuery`
- Something entirely else?

See, the killer feature here would be to make these queries `ResultQuery`, because they are in DB2, Firebird, PostgreSQL, and to some extent in Oracle, SQL Server. In DB2, they are even `Select`, or can behave like that. In PostgreSQL, they can be at least placed into WITH.

But for all the other RDBMS, letting these types extend `ResultQuery` would be a major distraction.

**TL;DR:**

Yes, we're aware of the inconvenience this part of our API currently poses. No, things are not as easy as they seem at first. Yes, we will fix it eventually. No we did not fix it yet. And no, I will not put a deadline to this functionality :)

I hope this helps.
